src/flamegraph/color/mod.rs

```diff
@@ -77,41 +72,83 @@ impl FromStr for Palette {
     }
 }
 
+struct NamehashVariables {
+    vector: f32,
+    weight: f32,
+    max: f32,
+    modulo: u8,
+}
+
+impl NamehashVariables {
+    fn init() -> Self {
+        NamehashVariables {
+            vector: 0.0,
+            weight: 1.0,
+            max: 1.0,
+            modulo: 10,
+        }
+    }
+
+    fn update(&mut self, character: u8) {
+        let i = f32::from(character % self.modulo);
+        self.vector += (i / f32::from(self.modulo - 1)) * self.weight;
+        self.modulo += 1;
+        self.max += self.weight;
+        self.weight *= 0.70;
+    }
+
+    fn result(&self) -> f32 {
+        (1.0 - self.vector / self.max)
+    }
+}
+
 /// Generate a vector hash for the name string, weighting early over
 /// later characters. We want to pick the same colors for function
 /// names across different flame graphs.
-fn namehash(name: &str) -> f32 {
-    let mut vector = 0.0;
-    let mut weight = 1.0;
-    let mut max = 1.0;
-    let mut modulo = 10;
-
-    let name = {
-        let looking_for_module_name = if name.starts_with('`') {
-            &name[1..]
-        } else {
-            name
-        };
-
-        if let Some(index) = looking_for_module_name.find('`') {
-            &looking_for_module_name[index + 1..]
-        } else {
-            name
+fn namehash<I: Iterator<Item = u8>>(mut name: I) -> f32 {
+    let mut namehash_variables = NamehashVariables::init();
+    let mut module_name_found = false;
+
+    // The original Perl regex is: $name =~ s/.(.*?)`//;
+    // Ie. we want to remove everything before the first '`'. If '`' is the first character,
+    // we remove everything before the second '`'. If there is no '`', we keep everything.
+    // This becomes tricky because we want to compute the hash and do the potential deletion
+    // ine one pass only.
+    // So, we start computing the hash and we check for '`' after the first character.
+    // If we find '`' before the end of the computation (3 characters), we stop the computation.
+    // If the computation finishes normally, we search for the first next '`'.
+    // After that, either we have found a '`' (end of prefix), and we need to compute the hash from there,
+    // or there is no '`' in the iterator and we have the hash computed!
+
+    match name.next() {
+        None => return namehash_variables.result(),
+        Some(first_char) => namehash_variables.update(first_char),
+    }
+
+    for character in name.by_ref() {
```

Comment:
I guess we could just do:
```rust
while let Some(character) = name.next() {
```
which is a little easier to follow (I wasn't familiar with `by_ref()`)

Reply:
Or alternatively `by_ref().take(3)` and get rid of the `modulo` check?

Reply:
I think I prefer removing the `modulo` check, but I admit it is not easy to immediately know how much the iterator has been consumed with `by_ref`...

Reply:
[Here](https://github.com/jonhoo/inferno/pull/32/commits/01dacc3086dfedad8083642cf4486d4aad186a07)